### DISCUSSION TITLE
Watch for File Changes

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#814] [Docs]
<meta sectionIndex="2">
<meta url="https://github.com/johnlindquist/kit/discussions/814">
<meta id="D_kwDOEu7MBc4AP9Tk">
<meta title="Watch for File Changes">
<meta section="Script Metadata">
<meta i="2">    
<meta path="docs/watch-for-file-changes">

## Watch a Single File

The `// Watch` metadata enables you to watch for changes to a file on your system.

```js
// Name: Speak File
// Watch: ~/speak.txt

import "@johnlindquist/kit"

let speakPath = home("speak.txt")

try {
  let content = await readFile(speakPath, "utf-8")
  if (content.length < 60) {
    // We don't want `say` to run too long 😅
    say(content)
  }
} catch (error) {
  log(error)
}
```

[Open speak-file in Script Kit](https://scriptkit.com/api/new?name=speak-file&url=https://gist.githubusercontent.com/johnlindquist/ec65920283c6ef66429a2331cdc81539/raw/98584e1ee1cb6b5f4f235a6873bdfcb709dfb953/speak-file.js")

## Watch a Directory

The `// Watch` metadata uses [Chokidar](https://www.npmjs.com/package/chokidar) under the hood, so it supports the same glob patterns.

```js
// Name: Download Log
// Watch: ~/Downloads/*

import "@johnlindquist/kit"

// These are optional and automatically set by the watcher
let filePath = await arg()
let event = await arg()

if (event === "add") {
  await appendFile(home("download.log"), filePath + "\n")
}
```

[Open download-log in Script Kit](https://scriptkit.com/api/new?name=download-log&url=https://gist.githubusercontent.com/johnlindquist/395ced3283e44c8ed2fea885104a1346/raw/e9b03ddfb0ae969d2b82e45c41ac8680ea2e686b/download-log.js")

**Comment by NachoVazquez** [comment]
**👏 praise**:  Hi @johnlindquist! Thanks for this tool. I love it.

**❔ question**: Do we need some special treatment for spaced words in our path?

For example: `// Watch: ~/Applications/JetBrains Toolbox/WebStorm.app/Contents/Info.plist` is not working for me.

Thanks again!

**Reply by johnlindquist** [reply]
Hmm, I just tested with the following and it worked fine 🤔

```js
// Name: Test Watch Spaced
// Watch: ~/Spaced Dir/file.txt

import "@johnlindquist/kit"

let filePath = await arg()
let event = await arg()

log({filePath, event})
```

Kit.app may need `Full Disk Access` or `Files and Folders` then add `~/Applications`

<img width="748" alt="CleanShot 2022-07-25 at 15 18 52@2x" src="https://user-images.githubusercontent.com/36073/180875838-9f0bee39-5712-4053-9b4b-258db0e175db.png">

Please let me know if either of those fixes it for you 🙏

**Reply by NachoVazquez** [reply]
Awesome! Thanks for the help. I'll let you know, but that's probably it because I was trying with a non-spaced file and had the same behavior.

**Reply by NachoVazquez** [reply]
praise: It is working 🥳 ! 

question: is it possible that the `notify` alert will not work with a watch script?

**Reply by NachoVazquez** [reply]
I think that notify is the one not working for me :)

**Reply by NachoVazquez** [reply]
Maybe more permissions 😄

**Reply by johnlindquist** [reply]
Awesome! Thanks for reporting back. Do you know which one specifically fixed it?

---

Just tested this and it worked fine:

```js
// Name: Test Watch Spaced
// Watch: /Applications/Revoice Pro.app/Contents/Info.plist

import "@johnlindquist/kit"

let filePath = await arg()
let event = await arg()

notify({
    title: event,
    message: filePath
})
```

<img width="342" alt="CleanShot 2022-07-25 at 15 34 52@2x" src="https://user-images.githubusercontent.com/36073/180877980-7b1fbf36-fcd4-4616-9649-7f639dcc959a.png">


You may need to tweak the notifications prefs:
<img width="749" alt="CleanShot 2022-07-25 at 15 36 17@2x" src="https://user-images.githubusercontent.com/36073/180878103-464c33a3-bd85-49e6-a0f5-ae63e74395c4.png">

**Reply by NachoVazquez** [reply]
I think it was working all the time, notifications are what is not working.

I think both permissions work just fine. I'll let you know if I can make notifications work I tried a similar script to the one you shared but the problem is that I don see `Kit` listed on my notification pref pane.

**Reply by NachoVazquez** [reply]
![Screen Shot 2022-07-25 at 18 43 51](https://user-images.githubusercontent.com/9338604/180879082-2ac67880-6257-46dc-aef1-3da75ed089b7.png)

This simple script just hangs and does nothing.

**Reply by NachoVazquez** [reply]
![Screen Shot 2022-07-25 at 18 45 02](https://user-images.githubusercontent.com/9338604/180879234-bbe648d3-1949-49a6-b672-b146a50302b5.png)

**Reply by johnlindquist** [reply]
Hmm, can you try running:

```bash
 ~/.kenv/bin/just-notify
```

in your terminal? This should at least confirm the SDK is working...

I'll have to investigate why OSX isn't detecting that Kit.app is sending notifications 🤔

---

The reason `just-notify` looks like it's "hanging" is because the process that triggers the notification stays active until the notification lifecycle is complete. It takes about ~10 seconds.

`notify` is a simple wrapper around [https://www.npmjs.com/package/node-notifier](https://www.npmjs.com/package/node-notifier), so it's waiting for "clicks" and other interactions while the notification is "running".

If you want to hide the prompt after a notify, use `hide()`

```js
notify("hello")
hide()
```

**Reply by johnlindquist** [reply]
Also, if you want to "mock" a notification, you can use widgets:

```js
// Name: Notify Widget

import "@johnlindquist/kit"

let w = await widget(md(`## Something happened!`), {
    height: 100
})

await wait(3000)

w.close()
```

[Open notify-widget in Script Kit](https://scriptkit.com/api/new?name=notify-widget&url=https://gist.githubusercontent.com/johnlindquist/8b50702a4234655eba9fb87ac5aefbed/raw/f13cc9e3ac0bc8c13fb909946be1285e7a3bad0c/notify-widget.js")

**Reply by NachoVazquez** [reply]
Well, I applied the old trick and restart my laptop. It works now 😆 . It was probably the OS notification service hanging.

However, all these tricks will be very handy!! Thank you!

**Reply by NachoVazquez** [reply]
I like the notify widget, I'll use it in my experiments!

Btw, the script I'm building will listen for updates in my Jetbrain Toolbox - WebStorm installation and update the terminal script with the right version.

The problem: The `ws` script needs to be in the system path, but the one that I'm using **/usr/local/bin** is not writable by JetBrain Toolbox.

There are simpler solutions, but I like to have my shell scripts in a single place. Therefore I will listen to any updates to WebStorm and update the script.

**Reply by johnlindquist** [reply]
Very cool. Glad a good old-fashioned restart kicked osx back into action 😅.

Let me know (in a new thread) if you have any other ideas/help. I'm always happy to explore new scripts.

Also, we're starting up a Discord server if you'd like to chat/lurk on Script Kit stuff :) https://discord.gg/q6cgGtzP
